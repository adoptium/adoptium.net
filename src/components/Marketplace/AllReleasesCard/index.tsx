"use client"

import React, { useState } from "react"
import Image from "next/image"
import { useTranslations, useLocale } from "next-intl"
import { BsCopy } from "react-icons/bs"
import { FiDownload } from "react-icons/fi"
import { FaRedo } from "react-icons/fa"
import ChecksumModal from "@/components/ChecksumModal"
import AnimatedPlaceholder from "@/components/AnimatedPlaceholder"
import type { MarketplaceRelease } from "@/hooks"

import { capitalize } from "@/utils/capitalize"
import { getImageForDistribution } from "@/hooks"
import { fetchExtension } from "@/utils/fetchExtension"

interface AllReleaseCardProps {
    results: MarketplaceRelease[] | null
    onReset?: () => void
}

const AllReleaseCard: React.FC<AllReleaseCardProps> = ({ results, onReset }) => {
    const t = useTranslations()
    const locale = useLocale();
    const [modalOpen, setModalOpen] = useState(false)
    const [currentChecksum, setCurrentChecksum] = useState("")

    interface OpenModalWithChecksum {
        (checksum: string): void
    }

    const openModalWithChecksum: OpenModalWithChecksum = (checksum) => {
        setCurrentChecksum(checksum)
        setModalOpen(true)
    }

    // Loading state with sleek animated placeholders
    if (results === null) {
        return (
            <div className="mt-8">
                <div className="grid gap-4">
                    {/* Create 3 skeleton cards for the loading state */}
                    {[1, 2, 3].map((index) => (
                        <div
                            key={index}
                            className="relative overflow-hidden rounded-xl bg-gradient-to-br from-[#200E46]/90 via-[#2B1A4F]/80 to-[#200E46]/90 backdrop-blur-sm border border-[#200E46]/50"
                        >
                            <div className="p-6">
                                <AnimatedPlaceholder>
                                    {/* Desktop Layout Skeleton */}
                                    <div className="hidden md:flex items-center justify-between">
                                        <div className="flex items-center flex-1">
                                            <div className="w-[140px] flex-shrink-0">
                                                <div className="h-5 bg-[#200E46]/60 rounded w-20 animate-pulse"></div>
                                            </div>
                                            <div className="w-[140px] flex-shrink-0 px-2">
                                                <div className="h-4 bg-[#200E46]/60 rounded w-24 animate-pulse"></div>
                                            </div>
                                            <div className="w-[140px] flex-shrink-0 flex justify-center px-2">
                                                <div className="w-20 h-14 bg-[#200E46]/60 rounded-xl animate-pulse"></div>
                                            </div>
                                            <div className="w-[120px] flex-shrink-0 px-2 space-y-1">
                                                <div className="h-3 bg-[#200E46]/60 rounded w-16 animate-pulse"></div>
                                                <div className="h-4 bg-[#200E46]/60 rounded w-20 animate-pulse"></div>
                                            </div>
                                            <div className="w-[100px] flex-shrink-0 px-2 space-y-1">
                                                <div className="h-3 bg-[#200E46]/60 rounded w-8 animate-pulse"></div>
                                                <div className="h-4 bg-[#200E46]/60 rounded w-16 animate-pulse"></div>
                                            </div>
                                            <div className="w-[100px] flex-shrink-0 px-2 space-y-1">
                                                <div className="h-3 bg-[#200E46]/60 rounded w-8 animate-pulse"></div>
                                                <div className="h-4 bg-[#200E46]/60 rounded w-12 animate-pulse"></div>
                                            </div>
                                        </div>
                                        <div className="flex items-center gap-3 flex-shrink-0 ml-4">
                                            <div className="h-10 bg-[#200E46]/60 rounded-lg w-24 animate-pulse"></div>
                                            <div className="h-10 bg-[#200E46]/60 rounded-lg w-32 animate-pulse"></div>
                                        </div>
                                    </div>

                                    {/* Mobile Layout Skeleton */}
                                    <div className="block md:hidden space-y-4">
                                        <div className="flex items-center justify-between">
                                            <div className="space-y-2">
                                                <div className="h-5 bg-[#200E46]/60 rounded w-24 animate-pulse"></div>
                                                <div className="h-4 bg-[#200E46]/60 rounded w-20 animate-pulse"></div>
                                            </div>
                                            <div className="w-24 h-16 bg-[#200E46]/60 rounded-xl animate-pulse"></div>
                                        </div>
                                        <div className="grid grid-cols-2 gap-3">
                                            {[1, 2, 3, 4].map((i) => (
                                                <div key={i} className="space-y-1">
                                                    <div className="h-3 bg-[#200E46]/60 rounded w-16 animate-pulse"></div>
                                                    <div className="h-4 bg-[#200E46]/60 rounded w-20 animate-pulse"></div>
                                                </div>
                                            ))}
                                        </div>
                                        <div className="flex gap-3 pt-2">
                                            <div className="flex-1 h-10 bg-[#200E46]/60 rounded-lg animate-pulse"></div>
                                            <div className="flex-1 h-10 bg-[#200E46]/60 rounded-lg animate-pulse"></div>
                                        </div>
                                    </div>
                                </AnimatedPlaceholder>
                            </div>
                        </div>
                    ))}
                </div>
            </div>
        )
    }

    // No results state with reset option
    if (results && results.length === 0) {
        return (
            <div className="mt-8">
                <div className="flex justify-center items-center min-h-[400px]">
                    <div className="text-center space-y-6 max-w-md">
                        <div className="w-24 h-24 mx-auto bg-gradient-to-br from-[#2B1A4F] to-[#3E3355] rounded-full flex items-center justify-center">
                            <svg
                                className="w-12 h-12 text-gray-400"
                                fill="none"
                                stroke="currentColor"
                                viewBox="0 0 24 24"
                            >
                                <path
                                    strokeLinecap="round"
                                    strokeLinejoin="round"
                                    strokeWidth={1.5}
                                    d="M9.75 9.75l4.5 4.5m0-4.5l-4.5 4.5M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                                />
                            </svg>
                        </div>
                        <div className="space-y-2">
                            <h3 className="text-2xl font-semibold text-white">No distributions found</h3>
                            <p className="text-gray-400 leading-relaxed">
                                No Java distributions match your current filter criteria. Try adjusting your selections or reset all filters to see available distributions.
                            </p>
                        </div>
                        {onReset && (
                            <button
                                onClick={onReset}
                                className="inline-flex items-center gap-2 px-6 py-3 bg-gradient-to-r from-[#FF1464] to-[#FF1464]/90 hover:from-[#FF1464]/90 hover:to-[#FF1464] text-white font-semibold rounded-full transition-all duration-300 transform hover:scale-105 shadow-lg hover:shadow-xl shadow-[#FF1464]/25"
                            >
                                <FaRedo className="w-4 h-4" />
                                Reset Filters
                            </button>
                        )}
                    </div>
                </div>
            </div>
        )
    }

    return (
        <div className="mt-8">
            <ChecksumModal open={modalOpen} setOpen={setModalOpen} checksum={currentChecksum} />

            {/* Modern Grid Layout */}
            <div className="grid gap-4">
                {results.map((release, index) => (
                    <div
                        key={index}
                        className="group relative overflow-hidden rounded-xl bg-gradient-to-br from-[#200E46]/90 via-[#2B1A4F]/80 to-[#200E46]/90 backdrop-blur-sm border border-[#200E46]/50 hover:border-[#FF1464]/40 transition-all duration-300 hover:shadow-2xl hover:shadow-[#FF1464]/10 hover:transform hover:scale-[1.02]"
                    >
                        {/* Subtle gradient overlay on hover */}
                        <div className="absolute inset-0 bg-gradient-to-r from-[#FF1464]/5 to-[#FF1464]/5 opacity-0 group-hover:opacity-100 transition-opacity duration-300" />

                        {/* Mobile Layout */}
                        <div className="block md:hidden space-y-4">
                            <div className="relative p-6">
                                {/* Header with version and vendor logo */}
                                <div className="flex items-center justify-between min-h-[4.5rem]"> {/* min-h for vertical centering */}
                                    <div className="flex flex-col justify-center">
                                        <h3 className="text-xl font-bold text-white mb-1">
                                            {release.release_name}
                                        </h3>
                                        <p className="text-sm text-slate-300">
                                            {capitalize(release.binary.distribution)}
                                        </p>
                                    </div>
                                    <div className="flex-shrink-0 flex items-center h-full">
                                        <div className="w-24 h-16 bg-white rounded-xl p-2.5 shadow-xl ring-2 ring-[#FF1464]/20 flex items-center justify-center">
                                            <Image
                                                className="w-full h-full object-contain drop-shadow-sm"
                                                alt={`${release.binary.distribution} logo`}
                                                src={getImageForDistribution(release.binary.distribution)}
                                                width={64}
                                                height={64}
                                            />
                                        </div>
                                    </div>
                                </div>

                                {/* Details Grid */}
                                <div className="grid grid-cols-2 gap-3 text-sm">
                                    <div className="space-y-1">
                                        <span className="text-slate-300">{t("Marketplace.Releases.build-date")}</span>
                                        <p className="text-white font-medium">
                                            {new Date(release.binary.timestamp).toLocaleDateString(locale, {
                                                day: "2-digit",
                                                month: "2-digit",
                                                year: "numeric",
                                            })}
                                        </p>
                                    </div>
                                    <div className="space-y-1">
                                        <span className="text-slate-300">OS</span>
                                        <p className="text-white font-medium">
                                            {capitalize(release.binary.os)}
                                        </p>
                                    </div>
                                    <div className="space-y-1">
                                        <span className="text-slate-300">{t("Temurin.Releases.ReleaseFilters.architecture")}</span>
                                        <p className="text-white font-medium">
                                            {release.binary.architecture}
                                        </p>
                                    </div>
                                    <div className="space-y-1">
                                        <span className="text-slate-300">{t("Marketplace.Releases.vendor")}</span>
                                        <p className="text-white font-medium">
                                            {release.vendor}
                                        </p>
                                    </div>
                                </div>

                                {/* Action Buttons - Grouped Design */}
                                <div className="bg-[#200E46]/60 backdrop-blur-sm rounded-xl p-3 border border-[#200E46]/40">
                                    <div className="flex gap-2">
                                        <button
                                            onClick={() => openModalWithChecksum(release.binary.package.sha256sum)}
                                            className="flex-1 flex items-center justify-center gap-2 px-4 py-2.5 bg-[#200E46]/80 hover:bg-[#200E46]/90 border border-[#FF1464]/20 hover:border-[#FF1464]/40 rounded-lg text-white text-sm font-medium transition-all duration-200 hover:transform hover:scale-105 hover:shadow-lg hover:shadow-[#FF1464]/20"
                                            aria-label="View checksum"
                                        >
                                            <BsCopy className="w-4 h-4" />
                                            {t("Temurin.Releases.ReleaseResults.checksum")}
                                        </button>
                                        <a
                                            href={typeof release.binary.package.link === 'string' ? release.binary.package.link : release.binary.package.link.toString()}
                                            className="flex-1 flex items-center justify-center gap-2 px-4 py-2.5 bg-gradient-to-r from-[#FF1464] to-[#FF1464]/90 hover:from-[#FF1464]/90 hover:to-[#FF1464] rounded-lg text-white text-sm font-semibold transition-all duration-200 shadow-lg hover:shadow-xl hover:shadow-[#FF1464]/30 hover:transform hover:scale-105"
                                        >
                                            <FiDownload className="w-4 h-4" />
                                            {t("Marketplace.Releases.download")} ({fetchExtension(release.binary.package.name)})
                                        </a>
                                    </div>
                                </div>
                            </div>
                        </div>

                        {/* Desktop Layout */}
                        <div className="hidden md:flex items-center justify-between min-h-[6rem] p-4">
                            <div className="flex items-center flex-1 min-h-[6rem]">
                                {/* Version */}
                                <div className="w-[140px] flex-shrink-0">
                                    <h3 className="text-lg font-bold text-white truncate">
                                        {release.release_name}
                                    </h3>
                                </div>

                                {/* Distribution */}
                                <div className="w-[140px] flex-shrink-0 px-2">
                                    <p className="text-white font-semibold truncate">
                                        {capitalize(release.binary.distribution)}
                                    </p>
                                </div>

                                {/* Vendor Logo - More Prominent */}
                                <div className="w-[140px] flex-shrink-0 flex justify-center px-2">
                                    <div className="w-20 h-14 bg-white rounded-xl p-2.5 shadow-xl ring-2 ring-[#FF1464]/20 hover:ring-[#FF1464]/40 transition-all duration-200">
                                        <Image
                                            className="w-full h-full object-contain drop-shadow-sm"
                                            alt={`${release.binary.distribution} logo`}
                                            src={getImageForDistribution(release.binary.distribution)}
                                            width={64}
                                            height={64}
                                        />
                                    </div>
                                </div>

                                {/* Build Date */}
                                <div className="w-[120px] flex-shrink-0 px-2">
                                    <span className="text-xs text-slate-300 block">{t("Marketplace.Releases.build-date")}</span>
                                    <p className="text-sm text-white font-medium truncate">
                                        {new Date(release.binary.timestamp).toLocaleDateString(locale, {
                                            day: "2-digit",
                                            month: "2-digit",
                                            year: "numeric",
                                        })}
                                    </p>
                                </div>

                                {/* OS */}
                                <div className="w-[100px] flex-shrink-0 px-2">
                                    <span className="text-xs text-slate-300 block">OS</span>
                                    <p className="text-sm text-white font-medium truncate">
                                        {capitalize(release.binary.os)}
                                    </p>
                                </div>

                                {/* Architecture */}
                                <div className="w-[100px] flex-shrink-0 px-2">
                                    <span className="text-xs text-slate-300 block">{t("Temurin.Releases.ReleaseFilters.architecture")}</span>
                                    <p className="text-sm text-white font-medium truncate">
                                        {release.binary.architecture}
                                    </p>
                                </div>
                            </div>

                            {/* Action Buttons - Grouped Design */}
                            <div className="bg-[#200E46]/40 backdrop-blur-sm rounded-xl p-3 border border-[#200E46]/30">
                                <div className="flex items-center gap-2">
                                    <button
                                        onClick={() => openModalWithChecksum(release.binary.package.sha256sum)}
                                        className="flex items-center gap-2 px-4 py-2 bg-[#200E46]/80 hover:bg-[#200E46]/90 border border-[#FF1464]/20 hover:border-[#FF1464]/40 rounded-lg text-white text-sm font-medium transition-all duration-200 hover:transform hover:scale-105 hover:shadow-lg hover:shadow-[#FF1464]/20"
                                        aria-label="View checksum"
                                    >
                                        <BsCopy className="w-4 h-4" />
                                        {t("Temurin.Releases.ReleaseResults.checksum")}
                                    </button>
                                    <a
                                        href={typeof release.binary.package.link === 'string' ? release.binary.package.link : release.binary.package.link.toString()}
                                        className="flex items-center gap-2 px-6 py-2 bg-gradient-to-r from-[#FF1464] to-[#FF1464]/90 hover:from-[#FF1464]/90 hover:to-[#FF1464] rounded-lg text-white text-sm font-semibold transition-all duration-200 shadow-lg hover:shadow-xl hover:shadow-[#FF1464]/30 hover:transform hover:scale-105"
                                    >
                                        <FiDownload className="w-4 h-4" />
                                        {t("Marketplace.Releases.download")} ({fetchExtension(release.binary.package.name)})
                                    </a>
                                </div>
                            </div>
                        </div>
                    </div>
                ))}
            </div>
        </div>
    )
}

export default AllReleaseCard
