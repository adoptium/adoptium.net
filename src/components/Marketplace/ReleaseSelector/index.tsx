"use client"

import React, { useMemo, useState } from "react"
import { useTranslations } from "next-intl"
import SelectorHeader from "@/components/Common/SelectorHeader"
import { useOses, useArches } from '@/hooks/fetchConstants'
import { packageTypes } from "@/utils/defaults"

interface ReleaseSelectorProps {
  marketplace?: boolean
  versions: {
    name: string
    value: string
  }[]
  defaultVersion?: string
  defaultOS?: string
  defaultArch?: string
  defaultPackageType?: string
  updateVersion: (version: string) => void
  updateOS: (os: string) => void
  updateArch: (arch: string) => void
  updatePackageType?: (pkgType: string) => void
}

const ReleaseSelector: React.FC<ReleaseSelectorProps> = ({
  marketplace,
  versions,
  defaultVersion,
  defaultOS,
  defaultArch,
  defaultPackageType,
  updateVersion,
  updateOS,
  updateArch,
  updatePackageType,
}) => {
  const t = useTranslations("Temurin.Releases.ReleaseFilters")
  // Track selected values for filtering
  const [selectedOS, setSelectedOS] = useState<string | undefined>(defaultOS);
  const [selectedArch, setSelectedArch] = useState<string | undefined>(defaultArch);

  // OS to architecture mapping for filtering
  const osToArchMap = useMemo(() => ({
    'windows': ['x64', 'x86', 'aarch64'],
    'mac': ['x64', 'aarch64'],
    'linux': ['x64', 'x86', 'arm', 'aarch64', 'ppc64le', 's390x', 'riscv64'],
    'alpine-linux': ['x64', 'aarch64'],
    'aix': ['ppc64'],
    'solaris': ['sparcv9', 'x64']
  }), []);

  // Architecture to OS mapping (derived from osToArchMap)
  const archToOsMap = useMemo(() => {
    const mapping: Record<string, string[]> = {};

    Object.entries(osToArchMap).forEach(([os, arches]) => {
      arches.forEach(arch => {
        if (!mapping[arch]) {
          mapping[arch] = [];
        }
        mapping[arch].push(os);
      });
    });

    return mapping;
  }, [osToArchMap]);

  // Custom wrappers for the update functions
  const handleOSChange = (os: string) => {
    setSelectedOS(os);
    updateOS(os);

    // If the current architecture isn't compatible with the new OS,
    // and the new OS isn't "any", reset architecture to "any"
    if (os !== 'any' && selectedArch && selectedArch !== 'any' &&
      !osToArchMap[os as keyof typeof osToArchMap]?.includes(selectedArch)) {
      setSelectedArch('any');
      updateArch('any');
    }
  };

  const handleArchChange = (arch: string) => {
    setSelectedArch(arch);
    updateArch(arch);

    // If the current OS isn't compatible with the new architecture,
    // and the new architecture isn't "any", reset OS to "any"
    if (arch !== 'any' && selectedOS && selectedOS !== 'any' &&
      !archToOsMap[arch]?.includes(selectedOS)) {
      setSelectedOS('any');
      updateOS('any');
    }
  };

  interface VersionNode {
    name: string;
    value: string;
  }

  const versionsList: VersionNode[] = versions.map((version: { name: string; value: string | number }): VersionNode => {
    return {
      name: version.name,
      value: String(version.value),
    }
  })

  // Get base lists
  const osList = useOses(true)
  const archList = useArches(true)

  // Filter OS list based on selected architecture
  const filteredOsList = useMemo(() => {
    if (osList.length === 0) return [];

    // If no architecture selected or "any" selected, show all OSes
    if (!selectedArch || selectedArch === 'any') {
      return osList;
    }

    // Otherwise, filter OSes to those compatible with the selected architecture
    return osList.filter(os => archToOsMap[selectedArch]?.includes(os.value));
  }, [osList, selectedArch, archToOsMap]);

  // Filter architecture list based on selected OS
  const filteredArchList = useMemo(() => {
    if (archList.length === 0) return [];

    // If no OS selected or "any" selected, show all architectures
    if (!selectedOS || selectedOS === 'any') {
      return archList;
    }

    // Otherwise, filter architectures to those compatible with the selected OS
    return archList.filter(arch => osToArchMap[selectedOS as keyof typeof osToArchMap]?.includes(arch.value));
  }, [archList, selectedOS, osToArchMap]);

  // Add "Any" option to the filtered lists
  const osListWithAny = useMemo(() => {
    if (filteredOsList.length === 0) return [];
    return [{ name: t("any-os"), value: "any" }, ...filteredOsList];
  }, [filteredOsList, t]);

  const archListWithAny = useMemo(() => {
    if (filteredArchList.length === 0) return [];
    return [{ name: t("any-architecture"), value: "any" }, ...filteredArchList];
  }, [filteredArchList, t]);

  // Create packageTypesWithAny unconditionally (following React Hooks rules)
  const packageTypesWithAny = useMemo(() => {
    return [{ name: t("any-package-type"), value: "any" }, ...packageTypes];
  }, [t]);

  // Conditionally use the data after hooks have been called
  const data = [osListWithAny, archListWithAny, versionsList];
  if (marketplace) {
    data.push(packageTypesWithAny);
  }

  const selectorUpdaters = [handleOSChange, handleArchChange, updateVersion]
  if (marketplace && updatePackageType) {
    selectorUpdaters.push(updatePackageType)
  }

  const titles = [t('operating-system'), t('architecture'), t('version'), t('package-type')]

  const defaultValues: string[] = []
  if (defaultOS) defaultValues.push(defaultOS)
  if (defaultArch) defaultValues.push(defaultArch)
  if (defaultVersion) defaultValues.push(defaultVersion)
  if (marketplace && defaultPackageType) {
    defaultValues.push(defaultPackageType)
  }

  return (
    <div>
      {" "}
      <div className="my-[50px]">
        <SelectorHeader
          data={data}
          selectorUpdater={selectorUpdaters}
          title={titles}
          defaultValues={defaultValues}
        />
      </div>
    </div>
  )
}

export default ReleaseSelector
