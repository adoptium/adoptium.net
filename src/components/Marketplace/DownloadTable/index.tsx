"use client"

import React, { useEffect, useState } from "react"
import ReleaseSelector from "@/components/Marketplace/ReleaseSelector"
import VendorSelector from "@/components/Marketplace/VendorSelector"
import AllReleaseCard from "@/components/Marketplace/AllReleasesCard"

import { setURLParam } from "@/utils/setURLParam"
import { detectOS, UserOS } from "@/utils/detectOS"
import { packageTypes, defaultArchitecture, defaultPackageType } from '@/utils/defaults'
import getVendorIdentifier from "@/utils/vendors"
import { getAllPkgsForVersion, MarketplaceRelease } from "@/hooks"
import { useArches, useOses } from '@/hooks/fetchConstants'
import vendors from "@/data/marketplace.json"
import { useSearchParams } from "next/navigation"
import { fetchAvailableReleases } from '@/utils/fetchAvailableReleases'

const DownloadTable = () => {
    const oses = useOses(true)
    const arches = useArches(true)
    const [ready, setReady] = useState(false)
    const searchParams = useSearchParams();

    const [os, updateOS] = useState("")
    const [arch, updateArch] = useState(defaultArchitecture)
    // Fix: always use string for version, and vendor identifiers
    const [version, updateVersion] = useState("21")
    const [packageType, updatePackageType] = useState(defaultPackageType)
    const [selectedVendorIdentifiers, updateSelectedVendorIdentifiers] = useState<string[]>([])
    const [releases, setReleases] = useState<MarketplaceRelease[] | null>(null)
    const [versions, setVersions] = useState<{ name: string; value: string }[]>([])

    // Parse query params from Next.js router
    useEffect(() => {
        if (oses.length === 0 || arches.length === 0) return;
        const params = Object.fromEntries(searchParams.entries());
        let defaultSelectedArch = defaultArchitecture;
        let defaultSelectedOS = "";
        let defaultSelectedPackageType = defaultPackageType;
        let defaultSelectedVersion = "21";

        // Arch param
        if (params.arch && arches.find(a => a.value === params.arch)) {
            defaultSelectedArch = params.arch;
        }
        // OS param
        if (params.os && oses.find(os => os.value === params.os)) {
            defaultSelectedOS = params.os;
        } else {
            const userOS = detectOS();
            switch (userOS) {
                case UserOS.MAC:
                    defaultSelectedOS = "mac";
                    if (typeof window !== "undefined" && typeof document !== "undefined") {
                        const gl = document.createElement("canvas").getContext("webgl");
                        const ext = gl && gl.getExtension("WEBGL_debug_renderer_info");
                        const param = (ext && gl.getParameter(ext.UNMASKED_RENDERER_WEBGL)) || "";
                        if (param.match(/Apple/) && !param.match(/Apple GPU/)) {
                            defaultSelectedArch = "aarch64";
                        }
                    }
                    break;
                case UserOS.LINUX:
                case UserOS.UNIX:
                    defaultSelectedOS = "linux";
                    break;
                default:
                    defaultSelectedOS = "windows";
                    break;
            }
        }
        // Package type param
        if (params.package_type && packageTypes.find(p => p.value === params.package_type)) {
            defaultSelectedPackageType = params.package_type;
        }
        // Version param
        if (params.version && typeof params.version === "string") {
            defaultSelectedVersion = params.version;
        }
        updateOS(defaultSelectedOS);
        updateArch(defaultSelectedArch);
        updateVersion(defaultSelectedVersion);
        updatePackageType(defaultSelectedPackageType);
        setReady(true);
    }, [oses, arches, searchParams]);

    // Fetch available versions for ReleaseSelector
    useEffect(() => {
        fetchAvailableReleases().then(setVersions)
    }, [])

    useEffect(() => {
        if (!ready) return;
        (async () => {
            setReleases(
                await getAllPkgsForVersion(
                    Number(version), // fix: ensure number
                    os,
                    arch,
                    packageType,
                    selectedVendorIdentifiers,
                ),
            )
        })()
    }, [ready, version, os, arch, packageType, selectedVendorIdentifiers])

    const versionUpdater = (version: string) => {
        setURLParam("version", version)
        updateVersion(version)
    }
    const archUpdater = (arch: string) => {
        setURLParam("arch", arch)
        updateArch(arch)
    }
    const osUpdater = (os: string) => {
        setURLParam("os", os)
        updateOS(os)
    }
    const handleReset = () => {
        const defaultOS = ""
        const defaultArch = defaultArchitecture
        const defaultVersion = "21"
        const defaultPackage = defaultPackageType
        // Fix: getVendorIdentifier expects a vendor object
        const defaultVendors: string[] = vendors.map(vendor => getVendorIdentifier(vendor))
        setURLParam("os", defaultOS)
        setURLParam("arch", defaultArch)
        setURLParam("version", defaultVersion)
        setURLParam("package_type", defaultPackage)
        updateOS(defaultOS)
        updateArch(defaultArch)
        updateVersion(defaultVersion)
        updatePackageType(defaultPackage)
        updateSelectedVendorIdentifiers(defaultVendors)
    }
    if (!ready) return null;
    return (
        <div className="max-w-[1264px] mx-auto px-6 pb-20">
            <ReleaseSelector
                marketplace
                versions={versions} // fix: pass versions prop
                updateVersion={versionUpdater}
                defaultVersion={version}
                updateOS={osUpdater}
                defaultOS={os}
                updateArch={archUpdater}
                defaultArch={arch}
                updatePackageType={updatePackageType}
            />
            <VendorSelector
                selectedVendorIdentifiers={selectedVendorIdentifiers}
                setSelectedVendorIdentifiers={updateSelectedVendorIdentifiers}
            />
            <AllReleaseCard results={releases} onReset={handleReset} />
        </div>
    )
}

export default DownloadTable