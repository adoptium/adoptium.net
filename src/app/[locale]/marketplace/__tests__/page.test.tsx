import React from "react"
import { render, cleanup } from "@testing-library/react"
import { describe, expect, it, afterEach, vi } from "vitest"
import { axe } from "vitest-axe"
import Marketplace from "../MarketplacePageClient"

// Add afterEach cleanup to clear timers and unmount components
afterEach(() => {
  cleanup();
  // Clear all timers to avoid lingering async work
  vi.clearAllTimers();
})

const ltsVersions = [
  { name: "11", value: "11" },
  { name: "17", value: "17" },
  { name: "21", value: "21" },
]

describe("Marketplace page", () => {
  it("renders correctly", () => {
    const { container } = render(<Marketplace ltsVersions={ltsVersions} latestLTS={21} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("has no accessibility violations", async () => {
    const { container } = render(<Marketplace ltsVersions={ltsVersions} latestLTS={21} />)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
  })
})
