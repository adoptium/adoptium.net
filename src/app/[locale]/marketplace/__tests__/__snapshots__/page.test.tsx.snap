// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Marketplace page > renders correctly 1`] = `
<div>
  <div
    class="relative pt-20 pb-8 md:pb-16 md:pt-32 overflow-hidden"
  >
    <div
      class="w-full relative"
    >
      <div
        class="absolute z-[-1] w-[2396px] h-[1340px] left-[23px] md:left-[112px] top-[21px] md:top-[71px] bg-[#410170] shadow-[0_0_400px_rgba(65,1,112,1)] rounded-full blur-[400px]"
      />
      <div
        class="absolute z-[-1] w-[1487px] h-[893px] left-[120px] md:left-[676px] top-[60px] md:top-[195px] bg-[#B62987] shadow-[0_0_400px_rgba(182,41,135,1)] rounded-full blur-[400px]"
      />
      <div
        class="absolute z-[-1] w-[688px] h-[446px] left-[400px] md:left-[1131px] top-[318px] bg-[#FE8492] shadow-[0_0_400px_rgba(254,132,146,1)] rounded-full blur-[400px]"
      />
    </div>
    <div
      class="mx-auto max-w-[1048px] w-full px-6 lg:px-0 flex flex-col items-center justify-center"
    >
      <div
        class="self-stretch flex-col justify-center items-center gap-6 flex"
      >
        <div
          class="self-stretch flex-col justify-center items-center gap-4 flex"
        >
          <div
            class="justify-start items-center gap-3 inline-flex"
          >
            <svg
              fill="none"
              height="14"
              viewBox="0 0 14 14"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#FF1464"
                height="8"
                rx="2"
                width="8"
                x="3"
                y="3"
              />
              <rect
                height="11"
                rx="3.5"
                stroke="#FF1464"
                stroke-opacity="0.25"
                stroke-width="3"
                width="11"
                x="1.5"
                y="1.5"
              />
            </svg>
            <div
              class="text-rose-600 text-base font-bold leading-normal"
            >
              Text
            </div>
          </div>
          <div
            class="self-stretch text-center text-white text-[56px] lg:text-[80px] leading-[114.286%] md:leading-[120%] font-semibold"
          >
            Text
          </div>
        </div>
        <div
          class="self-stretch text-center text-lightgrey text-[20px] font-normal leading-[140%] undefined"
        >
          Text
        </div>
      </div>
    </div>
  </div>
  <div
    class="about-aqavit flex justify-center items-center px-6 pt-10 pb-24 md:py-24"
  >
    <div
      class="aqavit-card gap-4 md:gap-10 flex flex-col md:flex-row justify-between md:items-center max-w-[1000px] mx-auto p-[32px] md:p-[64px]"
    >
      <div
        class="max-w-[110px] "
      >
        <img
          alt="AQAvit Logo"
          class="w-full"
          data-nimg="1"
          decoding="async"
          height="110"
          loading="lazy"
          src="/images/projects/aqavit-project.svg"
          style="color: transparent;"
          width="110"
        />
      </div>
      <div
        class="flex flex-col md:gap-3"
      >
        <div
          class="!mt-0"
        >
          <h2
            class="text-4xl leading-[122%] md:text-5xl md:[116%] text-white font-semibold"
          >
            About Eclipse AQAvit
          </h2>
          <p
            class="text-grey mt-6 font-normal text-base"
          >
            The AQAvit open source test suite (Adoptium Quality Assurance) can be found here. There is also a blog post and brief presentation that explains what testing is run and how it fits into the overall delivery pipeline.
          </p>
        </div>
        <a
          href="/aqavit"
        >
          <button
            class="bg-transparent mt-10 border-2 border-pink-500/0 text-white text-base  leading-6 font-normal w-[186px] h-[48px] rounded-[80px] gradient-border mt-6"
          >
            Learn More
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
